docs/knowledgebase/runtime/benchmarking.md

```diff
@@ -0,0 +1,547 @@
+
+    [features]
+    # -- snip --
+    runtime-benchmarks = ["frame-benchmarking"]
+    ```
+
+2. Start writing our benchmarking code. The code can be placed either in a `mod` module inside the
+   `src/lib.rs`, or be put in a separate file and include the module back in `src/lib.rs`. We will
+   use the first approach in the following to create the basic structure.
+
+    ```rust
+    #[cfg(feature = "runtime-benchmarks")]
+    mod benchmarking {
+      use super::*;
+      use frame_benchmarking::{benchmarks, account, impl_benchmark_test_suite};
+      use frame_system::RawOrigin;
+
+      benchmarks!{
+        // Individual benchmark cases here
+      }
+
+      impl_benchmark_test_suite!(Pallet, crate::tests::new_test_ext(),
+        crate::tests::Test);
+    }
+    ```
+
+3. Each benchmark case has three sections, setup section, execution section, and optionally a
+   verification section at the end.
+
+    ```rust
+    benchmarks!{
+      extrinsic_name {
+        // setup initial state
+      }: {
+        // running the benchmark
+      }
+      verify {
+        // Using assert statements to verify the result, or check certain on-chain event has been emitted.
+      }
+    }
+    ```
+
+    An example in
+    [`example/src/lib.rs`](https://github.com/paritytech/substrate/blob/master/frame/example/src/lib.rs):
+
+    ```rust
+    // This will measure the execution time of `set_dummy` for b in [1..1000] range.
+    set_dummy_benchmark {
+      let b in 1 .. 1000;
+    }: {
+      set_dummy (RawOrigin::Root, b.into());
+    }
+    ```
+
+    The name of the benchmark case is `set_dummy_benchmark`. Here `b` is the input that passed into
+    the extrinsic `set_dummy`. `b` will be varied from 1 to 1,000 and run repeatedly for each value
+    of `b` according to the arguments we passed when running the benchmark command. We will get back
+    to this in the following.
+
+    The function `set_dummy` is called in the execution section, and we do not verify the result at
+    the end.
+
+    You will also see code like this:
+
+    ```rust
+    accumulate_dummy {
+      let b in 1 .. 1000;
+      let caller = account("caller", 0, 0);
+    }: _ (RawOrigin::Signed(caller), b.into())
+    ```
+
+    This is when the execution section is just a one-line code that call the extrinsic with the same
+    name as the benchmarking case name. Our `benchmark!` macro is smart enough to know that!
+
+4.
+    ```rust
+    benchmarks! {
+      // -- snip
+      impl_benchmark_test_suite!(Pallet, crate::tests::new_test_ext(),
+        crate::tests::Test);
+    }
+    ```
+
+    At the last line in the `benchmarks!` macro, `impl_benchmark_test_suite!` takes three input: the
+    module name, the function that generates a test genesis storage `new_text_ext()`, and a test
+    runtime we built when writing our test cases, and expand them into testing code.
```

Comment:
```suggestion
    runtime, and expand them into testing code. The test runtime can often be the same one used for testing. 
```

This is better, because some benchmarks might need special mock-ing.